### ISSUE TITLE
ttnn.mean allocates **256-1024x more memory** than needed, depending on tensor shape.

### ISSUE THREAD

**Issue by sdawle-TT** [tenstorrent/tt-metal#32546]
### Describe the bug

`ttnn.mean()` with L1 memory config has a **critical bug** that causes it to allocate **256-1024x more memory** than needed, depending on tensor shape. The bug triggers at a threshold of approximately **1MB input size**.
This bug is causing the VAD v2 demo to fail with OOM issues.

### Test Results Summary

| Shape | Input Size | Expected Output | L1 Allocation Request | Over-allocation Factor | Result |
|-------|-----------|-----------------|----------------------|----------------------|--------|
| `(1, 4, 2, 2)` | 0.00 MB | 0.00 MB | Normal | 1x | ✓ PASS |
| `(2, 8, 4, 4)` | 0.00 MB | 0.00 MB | Normal | 1x | ✓ PASS |
| `(10, 64, 8, 8)` | 0.08 MB | 0.01 MB | Normal | 1x | ✓ PASS |
| `(100, 256, 1, 2)` | 0.10 MB | 0.05 MB | Normal | 1x | ✓ PASS |
| **`(1000, 256, 1, 2)`** | **0.98 MB** | **0.49 MB** | **500 MB** | **512x input, 1024x output** | **✗ FAIL** |
| `(2000, 256, 1, 2)` | 1.95 MB | 0.98 MB | 1000 MB | 512x input, 1024x output | ✗ FAIL |
| `(5000, 256, 1, 2)` | 4.88 MB | 2.44 MB | 2500 MB | 512x input, 1024x output | ✗ FAIL |
| `(10000, 256, 1, 2)` | 9.77 MB | 4.88 MB | 5000 MB | 512x input, 1024x output | ✗ FAIL |
| `(20000, 256, 1, 2)` | 19.53 MB | 9.77 MB | 10000 MB (10 GB!) | 512x input, 1024x output | ✗ FAIL (even DRAM!) |
| `(100, 512, 1, 4)` | 0.39 MB | 0.10 MB | 100 MB | 256x input, 1024x output | ✗ FAIL |
| `(10000, 128, 1, 2)` | 4.88 MB | 2.44 MB | 2500 MB | 512x input, 1024x output | ✗ FAIL |

### Key Findings

1. **Threshold**: Bug triggers between 0.10 MB and 0.98 MB input size
   - Last passing: `(100, 256, 1, 2)` = 0.10 MB
   - First failing: `(1000, 256, 1, 2)` = 0.98 MB

2. **Consistent Multiplier**: The over-allocation factor is **exactly 512x the input size**
   - 0.98 MB input → 500 MB allocation (512x)
   - 1.95 MB input → 1000 MB allocation (512x)
   - 4.88 MB input → 2500 MB allocation (512x)
   - 9.77 MB input → 5000 MB allocation (512x)

3. **Output Multiplier**: Always **exactly 1024x the expected output size**
   - This is 2x the input multiplier, which makes sense since output is half the size of input (reduction on last dim)

4. **Shape Independence**: The bug depends on **total tensor size**, not specific dimensions
   - `(10000, 256, 1, 2)` = 9.77 MB → 5000 MB allocation
   - `(10000, 128, 1, 2)` = 4.88 MB → 2500 MB allocation (half the size, half the allocation)
   - `(100, 512, 1, 4)` = 0.39 MB → 100 MB allocation (256x factor, different last dim)

5. **DRAM Also Fails**: At 20K batch (19.53 MB input), even DRAM fails trying to allocate 10 GB!

**Most Likely Cause:**
Intermediate buffer calculation error
   - The reduce operation may be allocating space for:
     - Input buffer: 1x
     - Output buffer: 0.5x (after reduction)
     - **Intermediate buffers**: Should be ~1-2x, but is actually 512x
   - This suggests a stride, padding, or shape calculation bug in the L1 allocator

Added a deepwiki link as well which shows the actual implementation of the reduce function:
https://deepwiki.com/search/i-am-seeing-a-but-in-ttnnmean_65472c5b-d8cc-475c-b69f-7d468cfeee7c?mode=fast
 

### Steps to reproduce the issue

- Pull the latest tt-metal, build it, set env variables
- Run the attached test_ttnn_mean_isolated.py

[test_ttnn_mean_isolated.py](https://github.com/user-attachments/files/23588635/test_ttnn_mean_isolated.py)

### Expected behavior
The Intermediate buffers size should be ~1-2x, but is actually 512x.

### Please complete the following environment information

OS: Ubuntu 22.04
Device: N300

**Comment by bbradelTT** [comment]
@sdawle-TT the output of mean (and similar ops) is in tile layout, regardless of the input layout. Therefore the expected sizes need to factor that in. We should be able to add in untilize on the output and have the output tensor be in row major format, but it'll take some time.

For now, you should have the output of the generic reduce ops (mean, sum, min, max, std, and var) be put in DRAM.

Also, our hardware is optimized to work on tiles. At the very least you should consider changing the height dimension to be something other than 1. And potentially look at the width dimension. For reduce, we'd transpose to the last dimension (which is of size 2), I'm not sure about other ops.

**Comment by sdawle-TT** [comment]
@bbradelTT Thank you for your input, currently for VADv2 running ttnn.mean() on DRAM is at least 10X slower than running on L1.
Any other suggestions that I could try to speed it up?

**Comment by bbradelTT** [comment]
I would suggest swapping dimension 1 and 2. Then tiled layout should be 32x smaller and computation 32x times faster.
I don't know if you can do that for other parts of the model. They would benefit from this as well probably.

**Comment by bbradelTT** [comment]
@sdawle-TT instead of swaping dimensions, you should be able to change the view of the tensor to be 3D instead of 4D, e.g. 20000, 256, 1, 2 to 20000, 256, 2, perform the reduction, and then change the tensor to row major and then back to 4D. I'll try to come up with an example tomorrow.

**Comment by bbradelTT** [comment]
Looking at DEVICE FW DURATION [ns] for an example shape, [64, 256, 1, 2], that fits into L1 the base and reshaped tensor for DRAM and L1 output shows:
<google-sheets-html-origin></style>
OP CODE | Base DRAM | Base L1 | 3D DRAM | 3D L1
-- | -- | -- | -- | --
TilizeWithValPadding | 1418433 | 1419727 | 106251 | 107172
Reduce | 735889 | 424605 | 34126 | 17523
ReshapeDeviceOperation | 363954 | 266141 | 139980 | 136535

Note that I have keepdim = False to see the overhead of reshape afterwards. You would want to have keepdim=True to avoid that reshape and then do your own.

Looks like a lot of the cost is in tilizing, which might be improved upon if fused with the different ops. However, the number one optimization for performance that needs to be done is that you need to do a zero cost reshape to not have the second last dimension be 1.

Code used for perf measurement:
```
@pytest.mark.parametrize(
        "input_shape, dim, keepdim", [
((64, 256, 1, 2), -1, False),
            ]
        )
def test_mean(device, input_shape, dim, keepdim):
    torch.manual_seed(0)
    torch_input_tensor = torch.rand(input_shape, dtype=torch.float32)
    torch_output_tensor = torch.mean(torch_input_tensor, dim, keepdim)

    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.float32,
            device=device)
    input_tensor = ttnn.reshape(input_tensor, [input_shape[0], input_shape[1] * input_shape[2], input_shape[3]]) # Base cases have this line commented out

    output_tensor = ttnn.mean(input_tensor, dim=dim, keepdim=keepdim,
            memory_config=ttnn.L1_MEMORY_CONFIG # DRAM cases have this line commented out
            )
```

**Comment by sdawle-TT** [comment]
Hi @bbradelTT
Currently, I am using the dimension swapping (dim1 and dim2) that you have described earlier, that seemed to have solved the OOM issue and made the test to work, but did not improve the perf much.

**I just tested your new approach on a small module in vadv2 with the tensor 4d tensor size (Shape([10000, 256, 1, 2]) reduced to 3d (Shape([10000, 256, 2]) for ttnn.mean operation, did not see much perf gain when using DRAM_MEMORY_CONFIG. I still can't use L1_MEMORY_CONFIG it fails with below OOM issue** 

`1.43 Kbps /138.5% =   0.00 Mbps | Tx: 71.3 KB | 64.06 MB | 11.96 s2025-11-26 19:35:03.724 | critical |          Always | TT_FATAL: Out of Memory: Not enough space to allocate 163840000 B L1 buffer across 64 banks, where each bank needs to store 2560000 B, but bank size is only 1364160 B (assert.hpp:103)`

So, instead of using 4.88MB it is trying to assign 155MB.

**Comment by cmaryanTT** [comment]
Idea: split on the last dimension, then reshape to move the [2] dimension to a different dimension, to make this tile friendly, then just do a (A+B)/2 to calculate the mean, and reshape back. Reshape will be costly, but the memory size should remains small.

**Comment by fahmedchanTT** [comment]
@sdawle-TT Need to try @cmaryanTT 's suggestion.

**Comment by sdawle-TT** [comment]
@fahmedchanTT - Yes, I will revisit this next week based on current priorities.

**Comment by iwroszTT** [comment]
@ign-navaneethk you may have a look at the pool_2d operation, which implements the row-major storage and tilize pattern
ttnn/cpp/ttnn/operations/pool/generic/device/kernels/compute/compute_pool_2d.cpp
See also the attached md file with AI generated walkthrough over that pattern.

[pool_tilize_reduce_pattern.md](https://github.com/user-attachments/files/27396554/pool_tilize_reduce_pattern.md)

**Comment by bbradelTT** [comment]
Merged PR caused perf regression. Functionality needs to be reversed until that can be investigated.

**Comment by bbradelTT** [comment]
Needs to be worked on together with #46110 to improve perf of the existing approach and undo changes from PR https://github.com/tenstorrent/tt-metal/pull/46122
